### PR TITLE
FoundationEssentials: tweak FileManager utilities for Windows

### PR DESCRIPTION
**Pull request by compnerd**
Add a helper for time conversion and remove some functions which do not make sense on Windows - users are not identified by integral IDs but rather by SIDs. This helps reduce the error diagnostics when building FoundationEssentials on Windows.

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:44]: I think it might make more sense to wait to include these as part of a PR that uses them. `NSTimeIntervalSince1970` is already available as `Date.timeIntervalBetween1970AndReferenceDate` so I think we can just use that instead of `NSTimeIntervalSince1970`, and it doesn't look like we use `NSTimeIntervalSince1601` here right?

Depending on the usages, it might make sense to rename this to `timeIntervalSinceReferenceDate` to indicate exactly what this represents and I suspect it can likely just be `internal`, but being able to see where it's used would be quite helpful

Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:44]: I can try to pull out parts of the changes that I have that use this. I am using these two constants for converting between `FILETIME` and `TimeInterval`. I don't see any reason to not use `internal` for the second one. However, I am not sure `timeIntervalSinceReferenceDate` is a great name for the constant - perhaps `Date.timeIntervalBetween1601AndReferenceDate` would work?

Comment (jmschonfeld) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: Ah sorry, I was suggesting `timeIntervalSinceReferenceDate` as the property name on `FILETIME` instead of just `interval`, I think having an internal `Date.timeIntervalBetween1601AndReferenceDate` for the constant seems fine if it's needed for Windows

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:409]: nit: do you think it's better to write it like this or maybe as `try? path.withNTPathRepresentation { ... } ?? false` if you want, just to avoid some extra nesting - thoughts?

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:396]:
```suggestion
                return GetFileAttributesExW($0, GetFileExInfoStandard, &faAttributes)
```

Does that work, or does this return an optional?

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:459]: To reuse code, could all of this just be

```swift
fileManager.isWritableFile(atPath: parent) && fileManager.isWritableFile(atPath: path)
```

That'd be a lot simpler / more readable if I'm write that it's the same behavior

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:561]: Maybe we could still call `_writeFileAttributePrimitive(_:as:)` here just so we're consistent even if it doesn't really do anything (since we do that on Linux too)?

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:564]: Are these a TODO for coming back later (if so, we could leave a TODO comment) or are they just not relevant for Windows (if so, we could leave out the comment and wrap the keys in `#if !os(Windows)`

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:612]: `withNTPathRepresentation` already calls `GetFullPathNameW` right? If so, is there an equivalent of `strlen` we could use here instead that might be faster?

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:617]: `withUnsafeTemporaryAllocation` might be a better choice here than an array - also do we need the `+1`?

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:639]: Same comment here about using `_writeFileAttributePrimitive(_:as:)` and the commented out lined below

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:732]: If these attributes aren't supported on windows, I think we should wrap their declarations in `#if !os(Windows)` so that it's clear to clients at compile time

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Files.swift:736]: Should we just have a `timeIntervalSince1601` property on `Date` for Windows that is `internal`?

**Review by jmschonfeld — COMMENTED** [review]
Thank you so much for your work on this! Left some comments inline but I really appreciate you iterating on this to help us fix up Windows support!

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:409]: As long as I can get it to compile, I think I prefer the `try? ... ?? false` variant.

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:564]: There in lies the question - I would like to believe that we should define them, even on Windows. I would rather that we mark these as TODO - yes, some attributes are unavailable, but, hopefully the combination of unit tests and real world usage would expose those. This allows us to reduce the noise in the build and get to the point where we might be able to draw a line in the sand to avoid backsliding the platform support.

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:612]: Not in the case that you have a NT path:

```swift
return try path.withCString(encodedAs: UTF16.self) { pwszPath in
  guard !path.hasPrefix(#"\\"#) else { return try body(pwszPath) }
  ...
}
```

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:617]: Aye and nay.

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:732]: TBH, I don't know about these. I'm just replicating the non-Darwin path.

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:736]: I like that!

File: Sources/FoundationEssentials/Date.swift
Comment (parkera) [inline Sources/FoundationEssentials/Date.swift:126]: Can this be internal or package?

Comment (compnerd) [reply Sources/FoundationEssentials/Date.swift:126]: That is meant to be `internal`, copy-pasta :(.

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:345]: Yeah, I think that `(false, false)` is the right behaviour. If you have a path that either doesn't conform to the requirements (and thus cannot exist), it makes sense to report that as non-existent. Additionally, the API call failing is expected for the cases where the file doesn't exist yet.

**Comment by compnerd** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:541]: Ah gotcha ok yeah I see now. Marking it as TODO looks good then we can always decide which are applicable later once we get it building

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:536]: I suppose that should be harmless, but I find that it does hurt readability somewhat.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test Windows

**Comment by compnerd** [comment]
Let's see where we stand with this PR - I still have some backlog to work through so I cannot determine this very well locally.

**Comment by compnerd** [comment]
@parkera, @jmschonfeld I might need help to get this merged - I wanted to see how we were doing on Windows, and didn't realise that it would block the merge for the PR :(

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (jmschonfeld) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:345]: Looks like this one might still need to be resolved - we should swap out the `try!` for returning `(false, false)` right?

File: Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift
Comment (jmschonfeld) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: Do we want to leave this as just `interval`? What interval does it represent (1601 until the time, or 1970 until the time, or other?)

**Comment by jmschonfeld** [comment]
@compnerd I left two comments about things the look like they might still be unresolved, once we square those away we can help make sure this gets merged. Does that Windows run show us what you were looking for (that we're at least heading in the right direction w.r.t. the number of build failures)?

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift:345]: Oh, missed the `try!`

File: Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: I think so. The `interval` property converts the `FILETIME` to a `TimeInterval`. The `TimeInterval` initialiser uses the UNIX epoch, which is what this converts it to.

File: Sources/FoundationEssentials/FileManager/FileManager+Files.swift
Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Files.swift]: Done

**Comment by compnerd** [comment]
Yeah, it seems that we are headed in the right direction. We are getting further, but still have a long road.

**Comment by compnerd** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift
Comment (jmschonfeld) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: Gotcha, in that case do you think we should name this `timeIntervalSince1970` instead of just `interval` to make that clearer? My main worry is that `interval` or even `TimeInterval` on its own don't convey the "start" of the interval and it might make this more approachable/readable to specify the start of the interval (or perhaps just outright convert to a `Date` here if we always use this to represent an instance in time).

Also, am I correct in understanding that `FILETIME` represents the time since 1601? If so, shouldn't this be the following:

```swift
Double((self.dwHighDateTime << 32) | self.dwLowDateTime) - Date.timeIntervalBetween1601AndReferenceDate + Date.timeIntervalBetween1970AndReferenceDate
```

I'm not sure subtracting the interval between 1970 and the reference date from the interval between 1601 and the date in question get's us to the right value

Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: I suppose that I can name it `timeIntervalSince1970`. It is meant more as a type conversion. I guess the question is, do you think that this is better:

```swift
TimeInterval(ftTime)
```

or is this more obvious:

```
ftTime.interval
```

The calculation itself is correct, the following is from swift-corelibs-foundation:

```swift
extension TimeInterval {
  init(_ ftTime: FILETIME) {
    self = Double((ftTime.dwHighDateTime << 32) | ftTime.dwLowDateTime) - NSTimeIntervalSince1970;
  }
}
```

Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: BTW, a FileTime represents the number of 100-nanosecond intervals since the NT Epoch (1/1/1601).

Comment (parkera) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:36]: Keep in mind `TimeInterval` is simply a typealias of `Double`, so any extensions to it will also appear on any Double as well.

**Comment by compnerd** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift
Comment (jmschonfeld) [inline Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:39]: Thanks for updating this, I like the `timeIntervalSince1970` spelling a lot better (and it mirrors the `Date` API nicely). I think this is almost correct: if `count` is the interval between 1601 and the time in question, then subtracting `timeIntervalBetween1601AndReferenceDate` gets us the interval between the reference date (2001) and the time in question. I think we need to add `Date.timeIntervalBetween1970AndReferenceDate` here if we want this to be 1970-aligned (or we should rename this to `timeIntervalSinceReferenceDate` and use that `Date` initializer instead of we want this to be 2001-aligned)

Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift:39]: Ah, okay - that does end up being exactly what I had initially - except I don't remember why we ended up reworking this 🤔

Comment (compnerd) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift]: FWIW, I find it to be a bit verbose of a spelling and not very helpful, but 🤷‍♂️

**Comment by compnerd** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift
Comment (jmschonfeld) [reply Sources/FoundationEssentials/FileManager/FileManager+Utilities.swift]: Cool thanks for updating. Yeah it's a bit more verbose, but I think it'll be quite helpful for other developers reading this since it makes it very clear whether you should use this with `Date(timeIntervalSince1970:)` vs. `Date(timeIntervalSinceReferenceDate:)` for example. But this all looks good to me now

**Review by jmschonfeld — APPROVED** [review]
LGTM pending macOS/Linux CI just to make sure we didn't break anything there

**Pull request merged**
